File: apps/api/src/mcp/tools/reverse-dns.ts
import { z } from "zod";
import { toolRegistry } from "./instance";
import { callGoWorker } from "../../workers/go-client";

const input = z.object({
  ip: z.string().describe("IPv4 or IPv6 address to reverse-resolve"),
});

toolRegistry.register({
  name: "reverse_dns",
  description:
    "Live PTR lookup for an IP address — returns hostnames currently bound to it. Note: historical reverse-DNS (every name that ever pointed at this IP) requires a paid passive-DNS provider and is not available in the free-tier implementation.",
  inputSchema: input,
  costMillicredits: 1,
  handler: async (i, ctx) => {
    const res = await callGoWorker({
      requestId: crypto.randomUUID(),
      tenantId: ctx.tenantId,
      userId: ctx.userId,
      tool: "reverse_dns",
      input: i,
      timeoutMs: 10_000,
    });
    if (!res.ok) throw new Error(res.error?.message ?? "reverse_dns failed");
    return res.output;
  },
});

File: apps/go-worker/internal/tools/reverse_dns.go
package tools

import (
	"context"
	"errors"
	"net"
	"strings"
	"time"
)

type ReverseDNSOutput struct {
	IP        string   `json:"ip"`
	Hostnames []string `json:"hostnames"`
	TookMs    int64    `json:"tookMs"`
	Note      string   `json:"note,omitempty"`
}

// ReverseDNS performs a live PTR lookup. Returns hostnames bound to the IP today.
// Historical PTR (i.e. names that ever pointed at this IP) requires a paid passive-DNS
// provider — out of scope for the free-tier Phase-0 implementation.
func ReverseDNS(ctx context.Context, input map[string]any) (*ReverseDNSOutput, error) {
	ipStr, _ := input["ip"].(string)
	ipStr = strings.TrimSpace(ipStr)
	if ipStr == "" {
		return nil, errors.New("input.ip required")
	}
	if net.ParseIP(ipStr) == nil {
		return nil, errors.New("input.ip is not a valid IP address")
	}

	start := time.Now()
	r := &net.Resolver{PreferGo: true}
	names, err := r.LookupAddr(ctx, ipStr)
	out := &ReverseDNSOutput{
		IP:     ipStr,
		TookMs: time.Since(start).Milliseconds(),
		Note:   "live PTR only; historical reverse-DNS requires a paid passive-DNS provider",
	}
	if err != nil {
		// `no such host` (NXDOMAIN) is a normal "no PTR" answer, not a tool failure.
		var dnsErr *net.DNSError
		if errors.As(err, &dnsErr) && dnsErr.IsNotFound {
			return out, nil
		}
		return nil, err
	}
	for _, n := range names {
		out.Hostnames = append(out.Hostnames, strings.TrimSuffix(strings.ToLower(n), "."))
	}
	return out, nil
}
